lib/addons/velero/index.ts

```diff
@@ -0,0 +1,227 @@
+        }
+        else {
+            this.options = defaultProps
+        }
+    }
+
+    deploy(clusterInfo: ClusterInfo): void {
+        const cluster = clusterInfo.cluster;
+        const props = this.options;
+        
+        let bucketName: string; // AWS S3 Bucketname
+        let veleroNamespace: string; // K8s namespace that Velero get deployed onto
+       
+        // Create S3 bucket if no existing bucket, create s3 bucket and corresponding KMS key
+        if ( !props.values.configuration.backupStorageLocation.bucket ){
+             console.log("existing S3 Bucket does not exists, creating S3 bucket");
+             const bucket = new s3.Bucket(cluster, "velero-backup-bucket", {
+                encryption: s3.BucketEncryption.KMS_MANAGED, // Velero Known bug for support with S3 with SSE-KMS with CMK, thus it does not support S3 Bucket Key: https://github.com/vmware-tanzu/helm-charts/issues/83
+                blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL, // Block Public Access for S3
+                publicReadAccess: false,
+                enforceSSL: true // Encryption in Transit
+            });
+           
+            // Create S3 VPC Endpoint for the Velero pod to access S3 via VPC Endpoint instead of going to internet
+            cluster.vpc.addGatewayEndpoint("velero-backup-bucket-vpcEndPoint", {
+                service: GatewayVpcEndpointAwsService.S3
+            })
+            bucketName = bucket.bucketName;
+        }
+        else {
+            bucketName = props.values.configuration.backupStorageLocation.bucket
+        }
+
+        // Create Namespace if not specified
+        if (props.namespace){
+            // Create Namespace if the "create" option is false
+            if (props.namespace.create) {
+                console.log ("namespace:" + props.namespace.name + " does not existed, creating")
+                cluster.addManifest("velero-namespace",
+                {
+                    apiVersion: "v1",
+                    kind: "Namespace",
+                    metadata: { 
+                        name: props.namespace.name
+                    }
+                }
+                )
+                veleroNamespace = props.namespace.name;
+            }
+            else{
+                console.log ("namespace:" + props.namespace.name + " exists, not creating new namespace")
+                veleroNamespace = props.namespace.name;
+            }
+        }
+        else{
+            veleroNamespace = "velero"; // initial value of veleroNamespace
+        }
+
+        // Setup IAM Role for Service Accounts (IRSA) for the Velero Service Account
+        const veleroServiceAccount = cluster.addServiceAccount (
+            "velero-account",
+            {
+                name: "velero-account",
+                namespace: veleroNamespace
+            }
+        );
+        const s3bucket = s3.Bucket.fromBucketName(cluster, "S3Bucket", bucketName);
```

Comment:
We aspire to adhere to "Clean Code" conventions. My recommendation is to approach this "script style" method structurally and break into more manageable methods, for example:
```
this.createNamespaceIfNeeded( ...)
this.createServiceAccount(...)
this.createS3Bucket(...)
... run helmChart ...
```

This will at the very least make it easier to read and maintain.

Reply:
Totally agreed with you on the clean code conventions. But given that this line of code 
```typescript
const s3bucket = s3.Bucket.fromBucketName(cluster, "S3Bucket", bucketName);
```
is only to be passed to the iam resource name, and with the lines of code to create the s3 bucket, i am thinking to keep the code as it is instead of wrapping them into a function.

Reply:
I guess I was not clear in my comment. I was not referring to the specific line of code, more referring to general structure of the deploy method. Indeed, extracting just s3 bucket does not make sense on its own. Here is what I propose:

```
deploy(clusterInfo: ClusterInfo): Promise<Construct> {
    this.createNamespaceIfNeeded( ...)
    this.createServiceAccount(...)
    this.createS3Bucket(...)
... run helmChart ...
    return Promise.resolve(helmChartConstruct); 
}
```

Also please note the return value as it can be leveraged for dependency if any other add-on needs to depend on this one. Also if returned it becomes accessible in team constructs and post deployment steps (e.g. apply some golden state from S3 bucket on cluster post install).

Reply:
@shapirov103  i got what you meant. I will refer to how argocd addon was written and wrap that to a similar format. will get back to you shortly.

Reply:
push the latest change please help to review to see whether it works as you expected. @shapirov103